src/realm/deppart/byfield_gpu_impl.hpp

```diff
@@ -0,0 +1,268 @@
+      entry.bounds = parent_space.bounds;
+      parent_entries_vec = {entry};
+      parent_entries = span<SparsityMapEntry<N,T>>(parent_entries_vec.data(), 1);
+    } else {
+      parent_entries = parent_space.sparsity.impl()->get_entries();
+    }
+
+    RegionInstance parent_entries_instance = this->realm_malloc(parent_entries.size() * sizeof(SparsityMapEntry<N,T>), my_mem);
+
+    RegionInstance inst_counters_instance = this->realm_malloc((field_data.size()) * sizeof(uint32_t), my_mem);
+
+    RegionInstance inst_offsets_instance = this->realm_malloc((field_data.size() + 1) * sizeof(size_t), my_mem);
+
+    SparsityMapEntry<N, T>* d_parent_entries = reinterpret_cast<SparsityMapEntry<N,T>*>(AffineAccessor<char,1>(parent_entries_instance, 0).base);
+    CUDA_CHECK(cudaMemcpyAsync(d_parent_entries, parent_entries.data(), parent_entries.size() * sizeof(SparsityMapEntry<N,T>), cudaMemcpyHostToDevice, stream), stream);
+
+    //This is used for count + emit: first pass counts how many rectangles survive intersection, second pass uses the counter
+    // to figure out where to write each rectangle
+    uint32_t* d_inst_counters = reinterpret_cast<uint32_t*>(AffineAccessor<char,1>(inst_counters_instance, 0).base);
+    CUDA_CHECK(cudaMemsetAsync(d_inst_counters, 0, (field_data.size()) * sizeof(uint32_t), stream), stream);
+
+
+    //This is used to track which instance each rectangle came from
+    size_t* d_inst_offsets = reinterpret_cast<size_t*>(AffineAccessor<char,1>(inst_offsets_instance, 0).base);
+    CUDA_CHECK(cudaMemcpyAsync(d_inst_offsets, inst_offsets.data(), (field_data.size() + 1) * sizeof(size_t), cudaMemcpyHostToDevice, stream), stream);
+
+
+    //First pass: figure out how many rectangles survive intersection
+    int flattened_threads = 256;
+    int grid_size = (parent_entries.size() * inst_size + flattened_threads - 1) / flattened_threads;
+    intersect_input_rects<N,T><<<grid_size, flattened_threads, 0, stream>>>(d_inst_entries, d_parent_entries, d_inst_offsets, nullptr, inst_size, parent_entries.size(), field_data.size(), d_inst_counters, nullptr);
+    KERNEL_CHECK(stream);
+
+
+    //Prefix sum over instances (small enough to keep on host)
+    std::vector<uint32_t> h_inst_counters(field_data.size()+1);
+    h_inst_counters[0] = 0; // prefix sum starts at 0
+    CUDA_CHECK(cudaMemcpyAsync(h_inst_counters.data()+1, d_inst_counters, field_data.size() * sizeof(uint32_t), cudaMemcpyDeviceToHost, stream), stream);
+    CUDA_CHECK(cudaStreamSynchronize(stream), stream);
+    for (size_t i = 0; i < field_data.size(); ++i) {
+      h_inst_counters[i+1] += h_inst_counters[i];
+    }
+
+    size_t num_valid_rects = h_inst_counters[field_data.size()];
+
+    if (num_valid_rects == 0) {
+      for (std::pair<const FT, SparsityMap<N, T>> it : sparsity_outputs) {
+        SparsityMapImpl<N, T> *impl = SparsityMapImpl<N, T>::lookup(it.second);
+        impl->gpu_finalize();
+      }
+      inst_offsets_instance.destroy();
```

Comment:
I don't see some of these used in the `else` case until they are deleted later anyway. Can they just be unconditionally deleted?